### PR TITLE
Throwing my hat in the ring for how to handle pathnames.

### PR DESCRIPTION
**Pull request by indifferentghost**
This is a suggestion of how to handle pathnames in mock-url #6.

Originally I had the pathname as an input field, but if the [can-route-pushstate.root](https://canjs.com/doc/can-route-pushstate.prototype.root.html) was set it would _ever so often_ reset to the root.

Demo: https://codepen.io/TDillard/pen/WaPmKw?editors=0010

### CODE REVIEW

**Comment by indifferentghost** [comment]
It's now configurable!
```html
<mock-url pushstate:from="true"></mock-url>
```
Enables it to catch [pushstate](https://developer.mozilla.org/en-US/docs/Web/API/History_API#Example_of_pushState()_method).

File: mock-url.mjs [hunk 1]
```diff
@@ -1,98 +1,129 @@
-
-import { Component, DefineMap, stache } from "//unpkg.com/can@5/core.mjs";
+import { Component, DefineMap, stache } from "can";
```
Comment (phillipskevin) [inline mock-url.mjs]: Removing the url for `can` breaks the demo page.

File: mock-url.mjs [hunk 2]
```diff
@@ -1,98 +1,129 @@
...
-        </div>`,
+	tag: "mock-url",
+	view: `
+		<div class='location'>
+			<span class='back' on:click='back()'>&#x21E6;</span>
+			<span class='forward' on:click='forward()'>&#x21E8;</span>
+			<span class='reload' on:click='reload()'>&#8635;</span>
+			<div class="url">
+				<span class='base'>URL:{{page}}{{path}}</span>
+				<span class='input'><input value:bind="url" placeholder="#! The hash is empty"/></span>
```
Comment (phillipskevin) [inline mock-url.mjs]: This doesn't seem to work correctly if using pushstate. The `path` should be editable.

File: mock-url.mjs [hunk 3]
```diff
@@ -1,98 +1,129 @@
...
+			</div>
+		</div>
+	`,
+	ViewModel: DefineMap.extend("MockUrl",{
+		page: {
+			default: "/my-app.html"
+		},
+		pushstate: {
+			default: false
```
Comment (phillipskevin) [inline mock-url.mjs]: Give this a `type: "boolean"` so that `:raw` or `:from` bindings will both work

File: mock-url.mjs [hunk 4]
```diff
@@ -1,98 +1,129 @@
...
+				var pushState = window.history.pushState;
+				
+				// prevent history.pushState from refreshing the page
+				window.history.pushState = function(){
+					window.history.replaceState.apply(this, arguments);
+					prop.resolve(location.pathname);
+				};
+				
+				// creating default pathname
+				window.history.replaceState( null, null, '/' );
```
Comment (phillipskevin) [inline mock-url.mjs:114]: If using a `root` other than `/`, won't this cause the URL to navigate away from the app?

**Review by phillipskevin — DISMISSED** [review]
I couldn't get this to work with an app that is using pushstate. I'll take another look once some of the requested changes are fixed up.

File: mock-url.mjs (hunk 4 above)
Comment (indifferentghost) [reply mock-url.mjs:114]: It'll only navigate if the `location.pathname` is directly edited, as far as I've been able to find.

**Pull request merged**